src/helpers/pivot/spreadsheet_pivot/spreadsheet_pivot.ts

```diff
@@ -0,0 +1,388 @@
+  /**
+   * This array contains the data entries of the pivot. Each entry is an object
+   * that contains the values of the fields for a row.
+   */
+  private dataEntries: DataEntries = [];
+  private fields: PivotFields = {};
+  /**
+   * This array contains the keys of the fields. It is used to keep the order
+   * of the fields as they are in the range.
+   */
+  private fieldKeys: TechnicalName[] = [];
+  /**
+   * This object contains the pivot table structure. It is created from the
+   * data entries and the pivot definition.
+   */
+  private table: SpreadsheetPivotTable | undefined;
+  /**
+   * This error is set when the range is invalid. It is used to show an error
+   * message to the user.
+   */
+  private invalidRangeError: EvaluationError | Error | undefined;
+  /**
+   * This flag is used to know when the pivot needs to be reloaded. It's only
+   * used in the evaluation process. At the end of each cycle, the flag is set
+   * to true so the pivot is reloaded in the next cycle.
+   */
+  needsReevaluation: boolean = true;
+
+  constructor(custom: ModelConfig["custom"], params: SpreadsheetPivotParams) {
+    this.getters = params.getters;
+    this.coreDefinition = params.definition;
+  }
+
+  init(params: InitPivotParams = {}) {
+    if (!this._definition || params.reload) {
+      this.invalidRangeError = undefined;
+      if (this.coreDefinition.dataSet) {
+        const { zone, sheetId } = this.coreDefinition.dataSet;
+        const range = this.getters.getRangeFromZone(sheetId, zone);
+        try {
+          ({ fields: this.fields, fieldKeys: this.fieldKeys } = this.extractFieldsFromRange(range));
+        } catch (e) {
+          this.fields = {};
+          this.fieldKeys = [];
+          this.invalidRangeError = e;
+        }
+      } else {
+        this.invalidRangeError = new EvaluationError(
+          _t("The pivot cannot be created because the dataset is missing.")
+        );
+      }
+      this._definition = new SpreadsheetPivotRuntimeDefinition(
+        this.coreDefinition,
+        this.fields,
+        this.getters
+      );
+      this.table = undefined;
+      this.dataEntries = [];
+      const range = this._definition.range;
+      if (this.isValid() && range) {
+        this.dataEntries = this.extractDataEntriesFromRange(range);
+      }
+      this.needsReevaluation = false;
+    }
+  }
+
+  get isInvalidRange() {
+    return !!this.invalidRangeError;
+  }
+
+  get invalidRangeMessage() {
+    return this.invalidRangeError?.message;
+  }
+
+  get definition() {
+    if (!this._definition) {
+      throw new Error("Pivot not loaded yet");
+    }
+    return this._definition;
+  }
+
+  isValid(): boolean {
+    if (this.invalidRangeError || !this._definition) {
+      return false;
+    }
+    for (const measure of this.definition.measures) {
+      if (!measure.isValid) {
+        return false;
+      }
+    }
+    for (const dimension of this.definition.columns.concat(this.definition.rows)) {
```

Comment:
better to test this in 2 loops: concat allocates an array, and there is a change that looping over just 1 already fails